policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+
+    def resync(self):
+        """
+        Updates Calico store with new resource information
+        """
+        for namespace_key, ns in self.changed_namespaces.items():
+            # Create/update Namespace Profiles
+            if ns.create_ns_profile():
+                self.nsResourceVersion = ns.resourceVersion
+                self.namespaces[namespace_key] = ns
+                del self.changed_namespaces[namespace_key]
+
+        for service_key, svc in self.changed_services.items():
+            # Process Services
+            self.svcResourceVersion = svc.resourceVersion
+            self.services[service_key] = svc
+            del self.changed_services[service_key]
+
+        for pod_key, po in self.changed_pods.items():
+            # Apply Namespace policy to new/updated pods
+            if po.apply_ns_policy():
+                if po.remove_default_profile():
+                    self.poResourceVersion = po.resourceVersion
+                    self.pods[pod_key] = po
+                    del self.changed_pods[pod_key]
+
+        # As endpoint lists change, create/add new profiles to pods as
+        # necessary
+        # TODO: if some object not ready for resync, finish loop w/out moving to
+        # processed list
+        for ep_key, ep in self.changed_endpoints.items():
+            # Get policy of the Service and namespace. Uses same namespace/name
+            # dict key.
+            try:
+                svc_type = self.services[ep_key].type
+            except KeyError:
+                _log.warning("Service %s not yet in store" % ep_key)
+                continue
+
+            try:
+                ns_policy = self.namespaces[ep.namespace].policy
+            except KeyError:
+                _log.warning("Namespace %s not yet in store" % ep.namespace)
+                continue
+
+            _log.info("Using svc policy %s and ns policy %s" % \
+                      (svc_type, ns_policy))
+
+            if ns_policy == "Open" or svc_type == SVC_TYPE_NAMESPACE_IP:
+                # If namespace is open, or svc is closed, do nothing.
+                # Namespace_IP services can only exists in Closed namespaces,
+                # in which case NamespaceIP service policy is redundant
+                # When the Namespace is open, all valid service types are open as well.
+                self.endpoints[ep_key] = ep
+                del self.changed_endpoints[ep_key]
+            else:
+                # else define service policy
+                associations = ep.generate_svc_profiles_pods()
+                for profile, pods in associations.items():
+                    # find pods and append profiles
+                    for pod in pods:
+                        existing_pod = self.match_pod(
+                            namespace=ep.namespace, name=pod)
+                        _log.debug("Adding profile %s to pod %s" % (profile, existing_pod))
+                        if existing_pod:
```

Comment:
This could be encapsulated inside the pod object for code re-use.  See my comment above regarding the API, but it might look like this:
```
pod.append_profile(...)
pod.set_profiles(...)
pod.remove_profile(...)
```